### PR TITLE
net: tcp2: Add support for IPv6

### PR DESCRIPTION
**Pull request by ozhuraki**
In order to support IPv4/IPv6, work with packet through
net_pkt_ip_hdr_len(), net_pkt_ip_opts_len() which account
for IPv4/IPv6, IPv4 options and IPv6 extension headers.

### CODE REVIEW

**Comment by rlubos** [comment]
Is this PR ready? I wanted to test it with `echo_client`/`echo_server` with TCP2/IPv6 enabled, but both crash at startup with the following changes:
```
diff --git a/samples/net/sockets/echo_server/prj.conf b/samples/net/sockets/echo_server/prj.conf
index 81d913e90d..9bdae36376 100644
--- a/samples/net/sockets/echo_server/prj.conf
+++ b/samples/net/sockets/echo_server/prj.conf
@@ -1,9 +1,12 @@
+CONFIG_HEAP_MEM_POOL_SIZE=8192
+
 # Generic networking options
 CONFIG_NETWORKING=y
-CONFIG_NET_UDP=y
+CONFIG_NET_UDP=n
 CONFIG_NET_TCP=y
+CONFIG_NET_TCP2=y
 CONFIG_NET_IPV6=y
-CONFIG_NET_IPV4=y
+CONFIG_NET_IPV4=n
 CONFIG_NET_SOCKETS=y
 CONFIG_NET_SOCKETS_POSIX_NAMES=y
 CONFIG_POSIX_MAX_FDS=6
```

Tried with `qemu_x86`.

**Comment by ozhuraki** [comment]
@rlubos

Thanks for the input. Do you test with E1000 against host?

The ipv6 support is ready and works against TTCN-3 sanity check and the included tcp sample.

I did a smoke test with echo_server trough SLIP against host before submitting, worked for me.

At the moment I am checking problems on the echo_client side (which we hadn’t tested before), already see problems and try to address them.

I will re-test the echo_server with E1000 with your config against host.

**Comment by rlubos** [comment]
@ozhuraki For starters I've tested qemu vs qemu:
* `west build -b qemu_x86 -t server` on `echo_server`
* `west build -b qemu_x86 -t client` on `echo_client`

Eventually, I want to test it with wireless L2 (OpenThread/IPSP/raw 802.15.4), we should sort out the qemu first though.

**Comment by ozhuraki** [comment]
@rlubos

> For starters I've tested qemu vs qemu

OK, thanks. I just started to check the client side and discovered that qemu vs qemu had been dropped some time ago, the documentation on this is obsolete.

The possibility is to test qemu against echo-client/echo-server from net-tools, other host tools (nc/telnet), TTCN-3.

> Eventually, I want to test it with wireless L2 (OpenThread/IPSP/raw 802.15.4), we should sort out the qemu first though.

OK. Let me check regarding the echo_server.

**Comment by rlubos** [comment]
Just to clarify:

>  I just started to check the client side and discovered that qemu vs qemu had been dropped some time ago, the documentation on this is obsolete.

I'm quite surprised by this information. I've been using this setup for testing for quite some time now and faced no issues so far (old TCP implementation still runs). Are there plans to remove  support for this scenario, along with the following documentation section:
https://docs.zephyrproject.org/1.13.0/subsystems/networking/qemu_setup.html#network-connection-between-two-qemu-vms
?

**Comment by ozhuraki** [comment]
@rlubos

> I'm quite surprised by this information.

I was also surprised and missed that. I think we should support it, let’s check what’s missing and add it.

Regarding the echo_server. I did a smoke test with E1000 against host, works for me:

1.	echo_server configuration (I added a few overlays for troubleshooting with 5cd93a1)

```
server-debug() {
    export BUILD=${ZEPHYR_BASE}/samples/net/sockets/echo_server/build
    rm -rf $BUILD && mkdir -p $BUILD && cd $BUILD
    cmake -DBOARD=qemu_x86 \
	  -DCONFIG_NET_UDP=n \
	  -DCONFIG_NET_IPV4=n \
	  -DCONFIG_NET_TCP2=y \
	  -DCONFIG_HEAP_MEM_POOL_SIZE=8192 \
	  -DOVERLAY_CONFIG="overlay-e1000.conf;overlay-log.conf;overlay-debug.conf;overlay-max-stacks.conf" \
	  ${@} \
	  ..
    make
}
```

2.	Create TAP interface with net-tools:

```
sudo ./net-setup.sh --config zeth.conf
```

3.	Run the app
4.	Smoke test the connection:

```
telnet -6 2001:db8::2 4242
```
P.S. The echo_client is known not to work. We just started to test with it, I am investigating to get past the initial problems.

File: subsys/net/ip/tcp2.c [hunk 1]
```diff
@@ -965,6 +965,7 @@ int net_tcp_update_recv_wnd(struct net_context *context, s32_t delta)
 /* net context wants to queue data for the TCP connection */
 int net_tcp_queue_data(struct net_context *context, struct net_pkt *pkt)
 {
+	int irq_level = irq_lock();
```
Comment (jukkar) [inline subsys/net/ip/tcp2.c]: What do we want to protect here, can you add more explanation to commit msg? Locking irqs seems quite heavy thing to do.

**Review by jukkar — COMMENTED** [review]
Lot of these commit messages were not answering the "why" question i.e., why this commit does what it does.

File: subsys/net/ip/tcp2.c [hunk 2]
```diff
@@ -163,7 +163,7 @@ static char *tcp_endpoint_to_string(union tcp_endpoint *ep)
 	}
 #undef BUF_SIZE
 #undef NBUFS
-	return s;
+	return log_strdup(s);
```
Comment (jukkar) [inline subsys/net/ip/tcp2.c]: This cannot be done like this. The log_strdup() call needs to be in the LOG_DBG() call, it cannot be used separately.

**Review by jukkar — CHANGES_REQUESTED** [review]

File: subsys/net/ip/tcp2.c (hunk 1 above)
Comment (ozhuraki) [reply subsys/net/ip/tcp2.c]: @jukkar 
> can you add more explanation to commit msg? Locking irqs seems quite heavy thing to do.

The message about interrupt proteciton is refined, please take a look.

Comment (pfl) [reply subsys/net/ip/tcp2.c]: Will net_tcp_queue_data() actually be called from the interrupt context, or is the lock protecting simultaneous access to this function or the context?

**Review by pfl — CHANGES_REQUESTED** [review]
The rest looks ok, comment added.

File: subsys/net/ip/tcp2.c (hunk 1 above)
Comment (jukkar) [reply subsys/net/ip/tcp2.c]: Why mutex or semaphore is not enough here?

File: subsys/net/ip/tcp2.c (hunk 2 above)
Comment (ozhuraki) [reply subsys/net/ip/tcp2.c]: @jukkar

> The log_strdup() call needs to be in the LOG_DBG() call, it cannot be used separately.

There's no other use for these functions outside of LOG_*(), but, OK, fixed.

File: subsys/net/ip/tcp2.c (hunk 1 above)
Comment (ozhuraki) [reply subsys/net/ip/tcp2.c]: @jukkar @pfl 

You are right, a mutex is a better idea here, so droped irq locks and added mutex, please take a look.

**Review by jukkar — APPROVED** [review]
LGTM

**Review by pfl — APPROVED** [review]

**Pull request merged**